xarray/tests/test_dataarray.py

```diff
@@ -6186,6 +6186,35 @@ def test_isin(da):
     assert_equal(result, expected)
 
 
+def test_coarsen_keep_attrs():
+    _attrs = {"units": "test", "long_name": "testing"}
+
+    da = xr.DataArray(
+        np.linspace(0, 364, num=364),
+        dims="time",
+        coords={"time": pd.date_range("15/12/1999", periods=364)},
+        attrs=_attrs,
+    )
+
+    da2 = da.copy(deep=True)
+
+    # Test dropped attrs
+    dat = da.coarsen(time=3, boundary="trim").mean()
+    assert dat.attrs == {}
```

Comment:
Is this the only test that fails? Does it also fail with `with set_options(keep_attrs=False):`?

Reply:
Actually on reviewing @dcherian 's comments I agree with his approach; I expect my suggested test will fail in the same way without that change. Let me know if it's unclear and I can help clarify.

Reply:
Yes this is the only failing test. I will try @dcherian 's suggestions.

Reply:
Now the dropped attribute test is passing, but the test for keeping the attributes fails.

Reply:
sorry, this is really messy.

`kwargs` does not have `keep_attrs` set here 
https://github.com/pydata/xarray/blob/66259d1853b85590bfbf6640fdfb868843812312/xarray/core/rolling.py#L692-L694

Basically we need to pass `self.keep_attrs` down to `Variable.coarsen`.
I would add ``keep_attrs`` to the signature of `Variable.coarsen` and then pass `self.keep_attrs` in `_reduce_func`

Reply:
You'll need to do the same here too: 
https://github.com/pydata/xarray/blob/66259d1853b85590bfbf6640fdfb868843812312/xarray/core/rolling.py#L700-L707

Reply:
Okay I've done that. I will look for any other place that might still be missing this, as the test_dataarray still does not pass.

Reply:
I just realized that they way I did this would ignore any global attribute set on keep_attrs so that might make some tests fail. I will fix this tomorrow

Reply:
let's add that to `Coarsen.__init__` when assigning `self.keep_attrs`